pkg/controller/controller.go

```diff
@@ -197,56 +197,87 @@ func SetLatestApiVersionAnnotation(object metav1.Object) {
 
 func ApplyVolumeRequestOnVMISpec(vmiSpec *v1.VirtualMachineInstanceSpec, request *v1.VirtualMachineVolumeRequest) *v1.VirtualMachineInstanceSpec {
 	if request.AddVolumeOptions != nil {
-		alreadyAdded := false
-		for _, volume := range vmiSpec.Volumes {
-			if volume.Name == request.AddVolumeOptions.Name {
-				alreadyAdded = true
-				break
-			}
-		}
-
-		if !alreadyAdded {
-			newVolume := v1.Volume{
-				Name: request.AddVolumeOptions.Name,
-			}
+		vmiSpec = handleAddVolumeRequest(vmiSpec, request)
+	} else if request.RemoveVolumeOptions != nil {
+		vmiSpec = handleRemoveVolumeRequest(vmiSpec, request)
+	}
 
-			if request.AddVolumeOptions.VolumeSource.PersistentVolumeClaim != nil {
-				newVolume.VolumeSource.PersistentVolumeClaim = request.AddVolumeOptions.VolumeSource.PersistentVolumeClaim
-			} else if request.AddVolumeOptions.VolumeSource.DataVolume != nil {
+	return vmiSpec
+}
 
-				newVolume.VolumeSource.DataVolume = request.AddVolumeOptions.VolumeSource.DataVolume
-			}
+func handleAddVolumeRequest(vmiSpec *v1.VirtualMachineInstanceSpec, request *v1.VirtualMachineVolumeRequest) *v1.VirtualMachineInstanceSpec {
+	if !volumeAlreadyAdded(vmiSpec, request) {
+		vmiSpec = addVolumeToSpec(vmiSpec, request)
+		vmiSpec = addVolumeAsDisk(vmiSpec, request)
+	}
 
-			vmiSpec.Volumes = append(vmiSpec.Volumes, newVolume)
+	return vmiSpec
+}
 
-			if request.AddVolumeOptions.Disk != nil {
-				newDisk := request.AddVolumeOptions.Disk.DeepCopy()
-				newDisk.Name = request.AddVolumeOptions.Name
+func handleRemoveVolumeRequest(vmiSpec *v1.VirtualMachineInstanceSpec, request *v1.VirtualMachineVolumeRequest) *v1.VirtualMachineInstanceSpec {
+	vmiSpec = removeVolumeFromSpec(vmiSpec, request)
+	vmiSpec = removeVolumeAsDisk(vmiSpec, request)
+	return vmiSpec
+}
 
-				vmiSpec.Domain.Devices.Disks = append(vmiSpec.Domain.Devices.Disks, *newDisk)
-			}
+func volumeAlreadyAdded(vmiSpec *v1.VirtualMachineInstanceSpec, request *v1.VirtualMachineVolumeRequest) bool {
+	alreadyAdded := false
+	for _, volume := range vmiSpec.Volumes {
+		if volume.Name == request.AddVolumeOptions.Name {
```

Comment:
Is it enough to just compare the names? Can we have a case where the same name is used but soe of the volume parameters has changed (if not, then we are fine) ?